redis/server.cc

```diff
@@ -152,26 +153,57 @@ redis_server::connection::~connection() {
 
 future<> redis_server::connection::process()
 {
-    return do_until([this] {
-        return _read_buf.eof();
-    }, [this] {
-        return with_gate(_pending_requests_gate, [this] {
-            return process_request().then_wrapped([this] (auto f) {
-                try {
-                    f.get();
-                }
-                catch (redis_exception& e) {
-                    write_reply(e);
-                }
-                catch (std::exception& e) {
-                    write_reply(redis_exception { e.what() });
-                }
-                catch (...) {
-                    write_reply(redis_exception { "Unknown exception" });
-                }
+    auto reader = [this] {
+        return do_until([this] { return _read_buf.eof(); }, [this] {
+            return with_gate(_pending_requests_gate, [this] {
+                _parser.init();
+                return _read_buf.consume(_parser).then([this] {
+                    if (_parser.eof()) {
+                        return make_ready_future<>();
+                    }
+                    ++_server._stats._requests_serving;
+                    utils::latency_counter lc;
+                    lc.start();
+                    promise<redis_server::result> pr;
+                    auto fut = pr.get_future();
+                    (void)process_request_internal(std::move(_parser.get_request())).then_wrapped([this, pr = std::move(pr), lc = std::move(lc)] (auto&& fut) mutable {
+                        _server._stats._requests.mark(lc.stop().latency());
+                        if (lc.is_start()) {
+                            _server._stats._estimated_requests_latency.add(lc.latency(), _server._stats._requests.hist.count);
+                        }
+                        fut.forward_to(std::move(pr)); 
+                    });
+                    return _queue.push_eventually(std::move(fut));
+                });
             });
         });
-    }).finally([this] {
+    };
+    auto writer = [this] {
+        return do_until([this] { return _read_buf.eof() && _queue.empty(); }, [this] {
```

Comment:
Is this eof() && empty() condition safe?
I'm worried that the reader code may read a request, before pushing onto the queue a future for handling this request, it defers for some reason (?) - so the writer thinks it has nothing to wait for, and fails to write the last reply....
One safer alternative is to have a special "eof" (end-of-file) marker on _queue, and writer will keep reading until it reads this eof marker. If there is no easy way to designate one value for eof, a simple approach is to use std::optional<T> for the queue's items - and a disengaged optional will mark the eof.

Reply:
You are right. The last reply may be lost. As you said, to avoid that, we should not use the eof of connection as the `EOF of writer. I'll fix it. Thank you.